In [4]:
import pandas as pd
import numpy as np

# Assessing
Use the space below to explore `all_alpha_08.csv` and `all_alpha_18.csv` to answer the quiz questions below.
- number of samples in each dataset
- number of columns in each dataset
- duplicate rows in each dataset
- datatypes of columns
- features with missing values
- number of non-null unique values for features in each dataset
- what those unique values are and counts for each

In [5]:
fuel_08_df = pd.read_csv('all_alpha_08.csv')
fuel_18_df = pd.read_csv('all_alpha_18.csv')

In [6]:
print(fuel_08_df.shape)
print(fuel_18_df.shape)

(2404, 18)
(1611, 18)


There are 18 colummns and 1611 samples in the 2018 dataset


In [7]:
print(fuel_08_df.duplicated().sum())
print(fuel_18_df.duplicated().sum())

25
0


In [8]:
fuel_08_df.isnull().sum()

Model                     0
Displ                     0
Cyl                     199
Trans                   199
Drive                    93
Fuel                      0
Sales Area                0
Stnd                      0
Underhood ID              0
Veh Class                 0
Air Pollution Score       0
FE Calc Appr            199
City MPG                199
Hwy MPG                 199
Cmb MPG                 199
Unadj Cmb MPG           199
Greenhouse Gas Score    199
SmartWay                  0
dtype: int64

In [9]:
fuel_18_df.isnull().sum()

Model                   0
Displ                   2
Cyl                     2
Trans                   0
Drive                   0
Fuel                    0
Cert Region             0
Stnd                    0
Stnd Description        0
Underhood ID            0
Veh Class               0
Air Pollution Score     0
City MPG                0
Hwy MPG                 0
Cmb MPG                 0
Greenhouse Gas Score    0
SmartWay                0
Comb CO2                0
dtype: int64

In [10]:
fuel_08_df.dtypes

Model                    object
Displ                   float64
Cyl                      object
Trans                    object
Drive                    object
Fuel                     object
Sales Area               object
Stnd                     object
Underhood ID             object
Veh Class                object
Air Pollution Score      object
FE Calc Appr             object
City MPG                 object
Hwy MPG                  object
Cmb MPG                  object
Unadj Cmb MPG           float64
Greenhouse Gas Score     object
SmartWay                 object
dtype: object

In [11]:
fuel_18_df.dtypes

Model                    object
Displ                   float64
Cyl                     float64
Trans                    object
Drive                    object
Fuel                     object
Cert Region              object
Stnd                     object
Stnd Description         object
Underhood ID             object
Veh Class                object
Air Pollution Score       int64
City MPG                 object
Hwy MPG                  object
Cmb MPG                  object
Greenhouse Gas Score      int64
SmartWay                 object
Comb CO2                 object
dtype: object

In [12]:
fuel_08_df.nunique()

Model                   436
Displ                    47
Cyl                       8
Trans                    14
Drive                     2
Fuel                      5
Sales Area                3
Stnd                     12
Underhood ID            343
Veh Class                 9
Air Pollution Score      13
FE Calc Appr              2
City MPG                 39
Hwy MPG                  43
Cmb MPG                  38
Unadj Cmb MPG           721
Greenhouse Gas Score     20
SmartWay                  2
dtype: int64

In [13]:
fuel_18_df.nunique()

Model                   367
Displ                    36
Cyl                       7
Trans                    26
Drive                     2
Fuel                      5
Cert Region               2
Stnd                     19
Stnd Description         19
Underhood ID            230
Veh Class                 9
Air Pollution Score       6
City MPG                 58
Hwy MPG                  62
Cmb MPG                  57
Greenhouse Gas Score     10
SmartWay                  3
Comb CO2                299
dtype: int64

How do the Cyl columns in the 2008 and 2018 datasets differ?
- The differ in Data types, Format and Number of unique values

In [14]:
fuel_08_df.Fuel.unique()

array(['Gasoline', 'ethanol/gas', 'ethanol', 'diesel', 'CNG'],
      dtype=object)

In [15]:
fuel_18_df.Fuel.unique()

array(['Gasoline', 'Gasoline/Electricity', 'Diesel', 'Ethanol/Gas',
       'Electricity'], dtype=object)

# Cleaning Column labels


In [16]:
fuel_08_df.head(1)

Model  Displ      Cyl    Trans Drive      Fuel Sales Area Stnd  \
0  ACURA MDX    3.7  (6 cyl)  Auto-S5   4WD  Gasoline         CA   U2   

   Underhood ID Veh Class Air Pollution Score FE Calc Appr City MPG Hwy MPG  \
0  8HNXT03.7PKR       SUV                   7          Drv       15      20   

  Cmb MPG  Unadj Cmb MPG Greenhouse Gas Score SmartWay  
0      17        22.0527                    4       no

In [18]:
fuel_18_df.head(1)

Model  Displ  Cyl       Trans Drive      Fuel Cert Region    Stnd  \
0  ACURA RDX    3.5  6.0  SemiAuto-6   2WD  Gasoline          FA  T3B125   

         Stnd Description  Underhood ID  Veh Class  Air Pollution Score  \
0  Federal Tier 3 Bin 125  JHNXT03.5GV3  small SUV                    3   

  City MPG Hwy MPG Cmb MPG  Greenhouse Gas Score SmartWay Comb CO2  
0       20      28      23                     5       No      386

### Drop Extraneous Columns

In [19]:
# drop columns from 2008 dataset
fuel_08_df.drop(['Stnd', 'Underhood ID', 'FE Calc Appr', 'Unadj Cmb MPG'], axis=1, inplace=True)

# confirm changes
fuel_08_df.head(1)

Model  Displ      Cyl    Trans Drive      Fuel Sales Area Veh Class  \
0  ACURA MDX    3.7  (6 cyl)  Auto-S5   4WD  Gasoline         CA       SUV   

  Air Pollution Score City MPG Hwy MPG Cmb MPG Greenhouse Gas Score SmartWay  
0                   7       15      20      17                    4       no

In [20]:
# drop columns from 2008 dataset
fuel_18_df.drop(['Stnd', 'Stnd Description', 'Underhood ID', 'Comb CO2'], axis=1, inplace=True)

# confirm changes
fuel_18_df.head(1)

Model  Displ  Cyl       Trans Drive      Fuel Cert Region  Veh Class  \
0  ACURA RDX    3.5  6.0  SemiAuto-6   2WD  Gasoline          FA  small SUV   

   Air Pollution Score City MPG Hwy MPG Cmb MPG  Greenhouse Gas Score SmartWay  
0                    3       20      28      23                     5       No

### Rename Columns

In [21]:
# rename Sales Area to Cert Region in 2008 dataset
fuel_08_df.rename(columns={'Sales Area':'Cert Region'}, inplace=True)

# confirm changes
fuel_08_df.head(1)

Model  Displ      Cyl    Trans Drive      Fuel Cert Region Veh Class  \
0  ACURA MDX    3.7  (6 cyl)  Auto-S5   4WD  Gasoline          CA       SUV   

  Air Pollution Score City MPG Hwy MPG Cmb MPG Greenhouse Gas Score SmartWay  
0                   7       15      20      17                    4       no

In [22]:
# replace spaces with underscores and lowercase labels for 2008 dataset
fuel_08_df.rename(columns=lambda x: x.strip().lower().replace(" ", "_"), inplace=True) # Strip method removes spaces(by default) at the beginning and end

# confirm changes
fuel_08_df.head(1)

model  displ      cyl    trans drive      fuel cert_region veh_class  \
0  ACURA MDX    3.7  (6 cyl)  Auto-S5   4WD  Gasoline          CA       SUV   

  air_pollution_score city_mpg hwy_mpg cmb_mpg greenhouse_gas_score smartway  
0                   7       15      20      17                    4       no

In [23]:
# replace spaces with underscores and lowercase labels for 2008 dataset
fuel_18_df.rename(columns=lambda x: x.strip().lower().replace(" ", "_"), inplace=True)

# confirm changes
fuel_18_df.head(1)

model  displ  cyl       trans drive      fuel cert_region  veh_class  \
0  ACURA RDX    3.5  6.0  SemiAuto-6   2WD  Gasoline          FA  small SUV   

   air_pollution_score city_mpg hwy_mpg cmb_mpg  greenhouse_gas_score smartway  
0                    3       20      28      23                     5       No

In [24]:
# confirm column labels for 2008 and 2018 datasets are identical
fuel_08_df.columns == fuel_18_df.columns

array([ True,  True,  True,  True,  True,  True,  True,  True,  True,
        True,  True,  True,  True,  True])

In [27]:
# make sure they're all identical like this
(fuel_08_df.columns == fuel_18_df.columns).all()

True